File: events/2015/01/solution.go
package main

import (
	"yod_aoc/utils/go/aoc_utils"
	"fmt"
	"time"
	"encoding/json"
	"os"
)

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
func part1(input []string) int {
	finalFloor := 0
	directions := input[0]

	for i := 0; i < len(directions); i++ {
		direction := string(directions[i])
		if direction == "(" {
			finalFloor++
		} else if direction == ")" {
			finalFloor--
		}
	}

	return finalFloor
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
func part2(input []string) int {
	currentFloor := 0
	directions := input[0]

	for i := 0; i < len(directions); i++ {
		direction := string(directions[i])
		if direction == "(" {
			currentFloor++
		} else if direction == ")" {
			currentFloor--
		}

		// If we are in the basement, return the current index
		if currentFloor == -1 {
			return i + 1 // +1 as AoC is 1-indexed
		}
	}

	// If we never enter the basement, return -1
	return -1
}

// ----------------------------------------------------------------------------------------------------
// | Main Function
// ----------------------------------------------------------------------------------------------------
func main() {
	// Read the arguments
	args := os.Args[1:]
	var fileName string
	if len(args) > 0 {
		fileName = args[0]
	} else {
		fileName = "input.txt"
	}
	// Start the timer!
	startTime := time.Now()

	// Read the input to an array of strings
	path := aoc_utils.GetFilePath(2015, 1, fileName)
	content, err := aoc_utils.ReadInputAsArray(path);
	if err != nil {
		fmt.Println("Error reading input:", err)
		return
	}

	// Run the Parts
	p1Result := aoc_utils.MeasurePerformance(func() interface{} {
		return part1(content)
	})
	
	p2Result := aoc_utils.MeasurePerformance(func() interface{} {
		return part2(content)
	})

	// End the Timer!
	endTime := time.Now()
	duration := endTime.Sub(startTime)

	resultsForDay := aoc_utils.AOCDayResults{
		Year: 2015,
		Day: 1,
		Part1: p1Result,
		Part2: p2Result,
		Duration: duration.Nanoseconds(),
		Timestamp: aoc_utils.AOCTimestamp{
			Start: startTime.Format(time.RFC3339),
			End: endTime.Format(time.RFC3339),
		},
	}
	jsonResults, _ := json.Marshal(resultsForDay)
	fmt.Println(string(jsonResults))
}

File: events/2015/01/solution.ts
import { AOCDayResults } from "@repo/utils-ts/returnType";
import { readInputAsArray } from "@repo/utils-ts/readInput";
import { measurePerformance } from "@repo/utils-ts/measurePerformance";

// ----------------------------------------------------------------------------------------------------
// | Part 1 > What floor does Santa end up on?
// ----------------------------------------------------------------------------------------------------
function part1(input: string[]): number | undefined {
  // Use finalFloor to track the floor Santa is on
  let finalFloor = 0;

  // Get the directions from the input
  const directions = input[0];

  // Work through the directions and update the floor as we go
  for (var i = 0; i < directions.length; i++) {
    const direction = directions[i];

    if (direction === '(') {
      finalFloor += 1;
    } else if (direction === ')') {
      finalFloor -= 1;
    }
  }

  // Return the final floor
  return finalFloor;
}

// ----------------------------------------------------------------------------------------------------
// | Part 2 > When does Santa first enter the basement?
// ----------------------------------------------------------------------------------------------------
function part2(input: string[]): number | undefined {
  // Keep track of the floor we are currently on
  let currentFloor = 0;

  // Get the directions from the input
  const directions = input[0];

  // Work through the directions and update the floor as we go
  for (var i = 0; i < directions.length; i++) {
    const direction = directions[i];

    if (direction === '(') {
      currentFloor += 1;
    } else if (direction === ')') {
      currentFloor -= 1;
    }

    // If we are in the basement, return the current index
    if (currentFloor === -1) {
      return i + 1; // +1 as AoC is 1-indexed
    }
  }

  // If we never enter the basement, return undefined
  return undefined;
}


// ----------------------------------------------------------------------------------------------------
// | Solve the puzzle
// ----------------------------------------------------------------------------------------------------
function solve() {
  // Get the arguments
  const args = process.argv.slice(2);
  const fileName = args[0] || "input.txt";

  // Start the timer
  const startTimeStamp = new Date(Date.now()).toISOString();
  const startTime = performance.now();

  // Read the input as an array of strings
  const input = readInputAsArray(fileName);

 // Run the parts
 const p1Result = measurePerformance(() => part1(input));
 const p2Result = measurePerformance(() => part2(input));

 // End the timer
 const endTimeStamp = new Date(Date.now()).toISOString();
 const endTime = performance.now();
 const duration = endTime - startTime;

 // Return the results
 return {
  Year: 2015,
  Day: 1,
  Part1: p1Result,
  Part2: p2Result,
  Duration: Math.round(duration * 1000000), // Convert to nanoseconds
  Timestamp: {
    Start: new Date(startTimeStamp).toISOString(),
    End: new Date(endTimeStamp).toISOString(),
  },
 } as AOCDayResults;
  
}
const results = solve();
console.log(JSON.stringify(results, null, 2));